crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -108,9 +108,9 @@ pub struct DatabaseProvider<TX> {
 }
 
 impl<TX> DatabaseProvider<TX> {
-    /// Returns snapshot provider
-    pub fn snapshot_provider(&self) -> Option<Arc<SnapshotProvider>> {
-        self.snapshot_provider.clone()
+    /// Returns a snapshot provider reference
+    pub fn snapshot_provider(&self) -> Option<&Arc<SnapshotProvider>> {
```

Comment:
can just clone here instead of &Arc or use deref

Reply:
We discussed externally that implementing it this way is challenging due to lifetime dependencies in `prepare_snapshotter`. The method `SnapshotProvider::get_writer` returns a mutable reference to a `DashMap `entry type (`SnapshotProviderRW`), which also holds an `Arc<SnapshotProvider>` of the same provider. Attempting to clone or use dereference in this context leads to the error of returning an object with a reference to a local object.

It was further to rename `SnapshotProvider` to `SnapshotProviderInner` and redefine `SnapshotProvider`as an `Arc<SnapshotProviderInner>`, since the usage of `SnapshotProvider` is always accompanied by an `Arc` over it.